src/main/java/net/neoforged/neoforge/event/BlockEntityTypesValidBlocksEvent.java

```diff
@@ -0,0 +1,112 @@
+import net.neoforged.fml.ModLoader;
+import net.neoforged.fml.common.EventBusSubscriber;
+import net.neoforged.fml.event.IModBusEvent;
+import net.neoforged.fml.event.lifecycle.FMLCommonSetupEvent;
+import org.jetbrains.annotations.Nullable;
+
+public class BlockEntityTypesValidBlocksEvent extends Event implements IModBusEvent {
+    private final ResourceKey<BlockEntityType<?>> blockEntityTypeResourceKey;
+    private final BlockEntityType<?> blockEntityType;
+    private final Set<Block> currentValidBlocks;
+    @Nullable
+    private Class<?> baseClass = null;
+
+    protected BlockEntityTypesValidBlocksEvent(ResourceKey<BlockEntityType<?>> blockEntityTypeResourceKey, BlockEntityType<?> blockEntityType) {
+        this.blockEntityTypeResourceKey = blockEntityTypeResourceKey;
+        this.blockEntityType = blockEntityType;
+        this.currentValidBlocks = new HashSet<>(blockEntityType.getValidBlocks());
+    }
+
+    /**
+     * The {@link ResourceKey} that this current {@link BlockEntityType} was registered under.
+     */
+    public ResourceKey<BlockEntityType<?>> getBlockEntityTypeResourceKey() {
+        return blockEntityTypeResourceKey;
+    }
+
+    /**
+     * The current {@link BlockEntityType}.
+     */
+    public BlockEntityType<?> getBlockEntityType() {
+        return blockEntityType;
+    }
+
+    /**
+     * An immutable list of the list of valid blocks that will be assigned to replace {@link BlockEntityType#validBlocks}'s value.
+     */
+    public Set<Block> getCurrentValidBlocks() {
+        return Collections.unmodifiableSet(this.currentValidBlocks);
+    }
+
+    /**
+     * Will add the given block to the list of valid blocks we will assign to the current {@link BlockEntityType}.
+     * Make sure the given block is the same or derived from the highest common class of all the entries in valid blocks.
+     * <p></p>
+     * Example: If the valid blocks list has {@link StandingSignBlock} entry and {@link WallSignBlock} entry, the common class is SignBlock
+     * and the given block must be a {@link SignBlock} or has {@link SignBlock} as a child class in its hierarchy.
+     * 
+     * @param block The block to add as a valid block for the current {@link BlockEntityType}
+     */
+    public void addValidBlock(Block block) {
+        setCommonSuperClassForExistingValidBlocks();
+
+        if (this.baseClass == null || this.baseClass.isAssignableFrom(block.getClass())) {
+            this.currentValidBlocks.add(block);
+        } else {
+            throw new IllegalArgumentException("Given block " + block + " does not derive from existing valid block's common superclass of " + this.baseClass);
+        }
+    }
+
+    private void setCommonSuperClassForExistingValidBlocks() {
+        if (this.baseClass == null) {
+            for (Block existingBlock : this.currentValidBlocks) {
+                if (this.baseClass != null) {
+                    if (!existingBlock.getClass().isAssignableFrom(this.baseClass)) {
```

Comment:
Hmm, this seems off. If `existingBlock.getClass().isAssignableFrom(this.baseClass)`, then the existing block class is a superclass of the base class. For instance, if the two blocks are of types `A extends B` and `B`, by the logic that's currently here, the loop would first set the base class to `A`, then keep it at `A` even though `B` is the common super class because it would skip this if block. Is this missing an `else` block?

Reply:
Nope. i tested it myself. if it kept it at A, then for signs, it would not downgrade to SignBlock and instead, stay at WallBlock.

If baseClass is SignBlock and existingBlock is WallSignBlock, it is false.
![image](https://github.com/neoforged/NeoForge/assets/40846040/6d51327a-9f5b-4e52-870e-e870540c0362)

the reverse would be true which then would downgrade baseClass to SignBlock
![image](https://github.com/neoforged/NeoForge/assets/40846040/24626795-8f82-4aa5-ae17-8f55f9fffa73)

Reply:
Yes that's the issue. It's _currently_ keeping it at A if you iterate the other way -- since `SignBlock.isAssignableFrom(WallSignBlock)` is false, the if block is skipped and the type is kept at WallSignBlock when it ought to be SignBlock, but only if it iterates in that order.

Reply:
Thinking more, the check is backwards. Base class should be first and existing blocks second